### DISCUSSION TITLE
Next Steps?

### DISCUSSION THREAD

**Discussion by DotNetDublin** [BlazingSaddlesSolutions/TimeKeeper#50] [General]
As you may have seen @jonhilt has been busy working on implementing a version of the code which uses vertical slices which we have now merged into the main branch.

There was a lot of good discussion (#47) around this and that discussion can continue however now that we've merged this update into the main branch my mind turns to what the next steps in the project should be.

From a personal point of view I'm quite keen to see if I can follow @jonhilt's example and add another feature following the pattern he's set out.

I'm also conscious that one of our aims is to mimic a real world project which leads me to ask what would we be communicating to the client at this point.

To my mind the obvious core function of TimeKeeper is allowing users to add/edit time entries and view time entries already added. Would it therefore be an idea to share with the client a non functioning wireframe of how this page would look in order to get early feedback? This could simply be a razor component minus the wiring if that makes sense?

Looking forward to hearing people's thoughts.

**Comment by harperjohn** [comment]
Nice!!! Excellent job @jonhilt and @DotNetDublin !!!

I pulled the latest code, it built and ran first time!!! API alone, then API and WASM, then API and Server, and then API and WASM *and* Server!

What a great set of first steps!! The huge "initial get-the-ball-rolling effort by @DotNetDublin and the subsequent (also huge) effort of @jonhilt and @DotNetDublin  to iron out a bunch of early growing pain things! Very nice!!

Since this is a "next step" discussion, I will mention a few things that come to mind:

- Maybe the app bar should show the version (WASM or Server) and then maybe like the Mudblazor demo app, allow to switch between WASM and Server.
- I had both Client and Server open, and when I added an entry to one, it did not "show up" in the other one. Of course, we have not implemented this yet, so the only way to do the "refresh" was to "refresh" the whole page... And this gets us to the whole application state kind of thing... and this is exactly what Fusion addresses. I am game to give Fusion a test... but I am equally open to any option here... but I think we really need to address the "application state" topic. And I get the sense that "refreshing" the page is something that Single Page Apps have to directly and explicitly address. At this simple (somewhat "stateless" stage) the refresh does not seem to break anything... in fact it is currently the only way to refresh the entry list without adding a new entry. But I also get the sense that as time goes on, an "un-handled" page refresh could cause (possibly subtle) problems. 
- Might be nice to add logging... as we inevitably will be breaking things as we add things.
- Might be nice to add unit testing... so that break less things... :)

**Reply by DotNetDublin** [reply]
Taking those one by one

> Maybe the app bar should show the version (WASM or Server) and then maybe like the Mudblazor demo app, allow to switch between WASM and Server.

Agreed and nice and easy!

> I had both Client and Server open, and when I added an entry to one, it did not "show up" in the other one. Of course, we have not implemented this yet, so the only way to do the "refresh" was to "refresh" the whole page... And this gets us to the whole application state kind of thing... and this is exactly what Fusion addresses. I am game to give Fusion a test... but I am equally open to any option here... but I think we really need to address the "application state" topic.

I feel (and could be wrong) that in a real world scenario users would use one or the other and wouldn't be given then option to switch between the two especially in our project scenario. That said the time tracking application I use in work has a mobile app and I noticed that if I start tracking time on the mobile app it's reflected almost immediately in the web app.

If you wanted to create a "Fusion" branch and experiment then that's brilliant.

> Might be nice to add logging... as we inevitably will be breaking things as we add things.

Agreed and good to implement while the amount of code is relatively small so the same logging practices can be followed as we add new features..

> Might be nice to add unit testing... so that break less things... :)

Agreed and again good to implement now so similar tests can be added as we add new features.

**Reply by harperjohn** [reply]
Great! (all your responses sound great!! Thanks!)

And yes - I will look into creating a Fusion branch. I have been away from my "Fusion research" for about a month now because it seemed to be going through lots of growth/evolution/changes (it is a very new project... barely a year old, with lots of work done quite recently)

In the meantime how are we going to address application state?

I would like for us to have a simple/normal/currently-standard/etc. way of dealing with "application state" so that 1) we and the project have "something", and 2) I have something to compare/contrast with Fusion.

**Reply by DotNetDublin** [reply]
In terms of application state I think we would use option 2 or 3 from https://jonhilton.net/blazor-state-management/ depending on the particular scenario.

To keep the discussion contained I think when deciding next steps we should focus on the what and why and then when we pick the areas to look at next we can have separate discussions on the how,

**Comment by DotNetDublin** [comment]
I was just saying to @harperjohn that I think to keep the "next steps" discussion contained we should focus on the what and why and then when we've selected what to look at in the next week or two weeks we can have separate discussions on the how,

One mind trick we can do to help us decide is to imagine that we have a meeting with the client at the end of next week and only one person in the meeting from the client side will have a technical background.

What could we have ready by next Friday that would show progress but more importantly garner feedback that would help shape how we approach the application.

My own take is that as the core function of TimeKeeper is allowing users to record time entries could we demonstrate a mock-up or semi functional version of the time entry page?

On seeing this the client may say "Looks nice but I don't see where we can do X".
